pkg/client/typed/dynamic/client.go

```diff
@@ -0,0 +1,194 @@
+// dynamic client.
+type ResourceClient struct {
+	cl       *client.RESTClient
+	resource *unversioned.APIResource
+	ns       string
+}
+
+// namespace applies a namespace to the request if the configured
+// resource is a namespaced resource. Otherwise, it just returns the
+// passed in request.
+func (rc *ResourceClient) namespace(req *client.Request) *client.Request {
+	if rc.resource.Namespaced {
+		return req.Namespace(rc.ns)
+	}
+	return req
+}
+
+// List returns a list of objects for this resource.
+func (rc *ResourceClient) List(opts api.ListOptions) (*runtime.UnstructuredList, error) {
+	result := new(runtime.UnstructuredList)
+	err := rc.namespace(rc.cl.Get()).
+		Resource(rc.resource.Name).
+		VersionedParams(&opts, pCodec).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// Get gets the resource with the specified name.
+func (rc *ResourceClient) Get(name string) (*runtime.Unstructured, error) {
+	result := new(runtime.Unstructured)
+	err := rc.namespace(rc.cl.Get()).
+		Resource(rc.resource.Name).
+		Name(name).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// Delete deletes the resource with the specified name.
+func (rc *ResourceClient) Delete(name string) error {
+	return rc.namespace(rc.cl.Delete()).
+		Resource(rc.resource.Name).
+		Name(name).
+		Do().
+		Error()
+}
+
+// Create creates the provided resource.
+func (rc *ResourceClient) Create(obj *runtime.Unstructured) (*runtime.Unstructured, error) {
+	result := new(runtime.Unstructured)
+	err := rc.namespace(rc.cl.Post()).
+		Resource(rc.resource.Name).
+		Body(obj).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// Update updates the provided resource.
+func (rc *ResourceClient) Update(obj *runtime.Unstructured) (*runtime.Unstructured, error) {
+	result := new(runtime.Unstructured)
+	name, err := getObjectName(obj)
+	if err != nil {
+		return result, err
+	}
+	err = rc.namespace(rc.cl.Post()).
+		Resource(rc.resource.Name).
+		Name(name).
+		Body(obj).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// Watch returns a watch.Interface that watches the resource.
+func (rc *ResourceClient) Watch(opts api.ListOptions) (watch.Interface, error) {
+	return rc.namespace(rc.cl.Get()).
+		Prefix("watch").
+		Resource(rc.resource.Name).
+		VersionedParams(&opts, pCodec).
+		Watch()
+}
+
+// paramaterCodec is a codec converts an API object to query
+// parameters without trying to convert to the target version.
+type parameterCodec struct{}
+
+func (parameterCodec) EncodeParameters(obj runtime.Object, to unversioned.GroupVersion) (url.Values, error) {
+	return queryparams.Convert(obj)
+}
+
+func (parameterCodec) DecodeParameters(parameters url.Values, from unversioned.GroupVersion, into runtime.Object) error {
+	return errors.New("DecodeParameters not implemented on nilParameterCodec")
+}
+
+var pCodec runtime.ParameterCodec = parameterCodec{}
+
+func getObjectName(obj *runtime.Unstructured) (string, error) {
```

Comment:
On Status in decode, I'm surprised you need to do that - request.go
should call decodeInto on your codec to check for status.

Having a wrapper that lives in typed client and uses the returned GVK
to transform specific kinds into known API types would probably be the
short term recommendation.  Another alternative is to create a scheme
and import specifically unversioned status and the other common types
(we need this anyway) and do a delegation call into the scheme.